File: Wdtc/src/main/java/org/wdt/wdtc/WdtcMain.java
package org.wdt.wdtc;


import org.apache.log4j.Logger;
import org.wdt.utils.io.FileUtils;
import org.wdt.wdtc.auth.UsersList;
import org.wdt.wdtc.auth.yggdrasil.AuthlibInjector;
import org.wdt.wdtc.game.Launcher;
import org.wdt.wdtc.game.config.GameConfig;
import org.wdt.wdtc.manger.*;
import org.wdt.wdtc.ui.ErrorWindow;
import org.wdt.wdtc.utils.JavaUtils;
import org.wdt.wdtc.utils.ThreadUtils;
import org.wdt.wdtc.utils.WdtcLogger;

import java.io.IOException;

public class WdtcMain extends JavaFxUtils {
    private static final Logger logmaker = WdtcLogger.getLogger(WdtcMain.class);

    public static void main(String[] args) {
        try {
            TaskManger.ckeckVMConfig();
            ckeckJavaFX();
            logmaker.info(String.format("===== Wdtc - %s - =====", VMManger.getLauncherVersion()));
            logmaker.info(String.format("Java Version:%s", System.getProperty("java.version")));
            logmaker.info(String.format("Java VM Version:%s", System.getProperty("java.vm.name")));
            logmaker.info(String.format("Java Home:%s", System.getProperty("java.home")));
            logmaker.info(String.format("Wdtc Debug Mode:%b", VMManger.isDebug()));
            logmaker.info("Wdtc Config Path:" + FileManger.getWdtcConfig());
            logmaker.info("Setting File:" + FileManger.getSettingFile());
            logmaker.info("Here:" + GameFolderManger.getDefaultHere());
            TaskManger.runStartUpTask();
            GameFileManger.downloadVersionManifestJsonFileTask();
            removePreferredVersion();
            UsersList.printUserList();
            AuthlibInjector.updateAuthlibInjector();
            GameConfig.writeConfigJsonToAllVersion();
            ThreadUtils.startThread(() -> JavaUtils.main(getRegistryKey())).setName("Found Java");
            AppMain.main(args);
        } catch (Throwable e) {
            ErrorWindow.setErrorWin(e);
        }
    }

    public static void removePreferredVersion() throws IOException {
        SettingManger.Setting setting = SettingManger.getSetting();
        if (setting.getPreferredVersion() != null) {
            Launcher launcher = new Launcher(setting.getPreferredVersion());
            if (FileUtils.isFileNotExists(launcher.getVersionJson())) {
                setting.setPreferredVersion(null);
                SettingManger.putSettingToFile(setting);
            }
        }
    }

    public static String[] getRegistryKey() {
        return new String[]{"HKEY_LOCAL_MACHINE\\SOFTWARE\\JavaSoft\\Java Development Kit",
                "HKEY_LOCAL_MACHINE\\SOFTWARE\\JavaSoft\\Java Update",
                "HKEY_LOCAL_MACHINE\\SOFTWARE\\JavaSoft\\Java Web Start Caps",
                "HKEY_LOCAL_MACHINE\\SOFTWARE\\JavaSoft\\JDK"};
    }
}

File: ui/src/main/kotlin/org/wdt/wdtc/ui/WdtcMain.kt
@file:JvmName("WdtcMain")

package org.wdt.wdtc.ui

import kotlinx.coroutines.launch
import org.wdt.utils.io.isFileNotExists
import org.wdt.wdtc.core.impl.auth.accounts.yggdrasil.downloadAuthlibInjector
import org.wdt.wdtc.core.impl.download.game.DownloadVersionGameFile.Companion.downloadVersionManifestJsonFileTask
import org.wdt.wdtc.core.impl.manger.ckeckRunEnvironment
import org.wdt.wdtc.core.impl.manger.createNeedDirectories
import org.wdt.wdtc.core.impl.manger.removeConfigDirectory
import org.wdt.wdtc.core.impl.manger.runStartUpTask
import org.wdt.wdtc.core.openapi.auth.printUserList
import org.wdt.wdtc.core.openapi.game.printVersionList
import org.wdt.wdtc.core.openapi.manager.*
import org.wdt.wdtc.core.openapi.utils.*
import org.wdt.wdtc.ui.loader.downloadDependencies
import org.wdt.wdtc.ui.loader.loadJavaFXPatch
import org.wdt.wdtc.ui.window.tryCatching
import javafx.application.Application.launch as launchApp

suspend fun main(args: Array<String>) {
	val startTime = System.currentTimeMillis()
	tryCatching {
		ckeckRunEnvironment()
		createNeedDirectories()
		if (args.isNotEmpty()) {
			removeConfigDirectory(args.first() == "refresh")
		}
		ckeckJavaFX()
		logmaker.info("===== Wdtc - $launcherVersion =====")
		logmaker.info("Current System:${System.getProperty("os.name")}")
		logmaker.info("Current Platform:$currentSystem")
		logmaker.info("Kotlin Version:${KotlinVersion.CURRENT}")
		logmaker.info("Java Version:${System.getProperty("java.version")}")
		logmaker.info("Java VM Version:${System.getProperty("java.vm.name")}")
		logmaker.info("Java Home:${System.getProperty("java.home")}")
		logmaker.info("Wdtc Debug Mode:$isDebug")
		logmaker.info("Wdtc Application Type:$applicationType")
		logmaker.info("Wdtc Config Path:$wdtcConfig")
		logmaker.info("Setting File:$settingFile")
		logmaker.info("Here:$defaultHere")
		runOnIO {
			runStartUpTask()
			launch {
				removePreferredVersion()
				printUserList()
				printVersionList()
			}
			if (isOnline) {
				launch {
					downloadVersionManifestJsonFileTask()
					downloadAuthlibInjector()
				}
			}
			launch(newSingleThreadContext()) { JavaUtils.main(registryKey) }
			launchApp(AppMain::class.java)
		}
		awaitApplicationBreak(startTime)
	}
}

private fun removePreferredVersion() {
	currentSetting.preferredVersion?.let {
		if (it.versionJson.isFileNotExists()) {
			currentSetting.saveSettingToFile {
				preferredVersion = null
			}
		}
	}
}

private fun awaitApplicationBreak(time: Long) {
	logmaker.info("Runtime: ${System.currentTimeMillis() - time}ms")
	logmaker.info("======= Wdtc Stop ========")
}

private val registryKey: Array<String> = arrayOf(
	"HKEY_LOCAL_MACHINE\\SOFTWARE\\JavaSoft\\Java Development Kit",
	"HKEY_LOCAL_MACHINE\\SOFTWARE\\JavaSoft\\Java Update",
	"HKEY_LOCAL_MACHINE\\SOFTWARE\\JavaSoft\\Java Web Start Caps",
	"HKEY_LOCAL_MACHINE\\SOFTWARE\\JavaSoft\\JDK"
)

suspend fun ckeckJavaFX() {
	try {
		Class.forName("javafx.application.Application")
	} catch (_: ClassNotFoundException) {
		downloadDependencies()
		loadJavaFXPatch()
	}
}

